File: app/src/main/java/com/imdglobalservices/psi/fragment/PSIStatisticFragment.java
package com.imdglobalservices.psi.fragment;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.formatter.IAxisValueFormatter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
import com.github.mikephil.charting.listener.OnChartValueSelectedListener;
import com.imdglobalservices.psi.R;
import com.imdglobalservices.psi.database.PSIDatabaseImplementation;
import com.imdglobalservices.psi.models.DataIndex;
import com.imdglobalservices.psi.models.PSIDate;
import com.imdglobalservices.psi.utils.ContentHelper;

import java.util.ArrayList;

/**
 * @author Dwi Setiyono <dwi.setiyono@ovo.id>
 * @since 2017.16.11
 */
public class PSIStatisticFragment extends Fragment implements OnChartValueSelectedListener {
    private LineChart lineChart;

    private ArrayList<String> arrayTime = new ArrayList<>();
    private ArrayList<Entry> arrayCentral = new ArrayList<>();
    private ArrayList<Entry> arrayNorth = new ArrayList<>();
    private ArrayList<Entry> arrayWest = new ArrayList<>();
    private ArrayList<Entry> arraySouth = new ArrayList<>();
    private ArrayList<Entry> arrayEast = new ArrayList<>();
    private ArrayList<Entry> arrayNational = new ArrayList<>();
    private ArrayList<ILineDataSet> dataSets = new ArrayList<>();
    private PSIDate psiByDate;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_statistic_psi, container, false);
        initView(view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        initLineChart();
        setData();
    }

    /**
     * this method for initialize all view
     *
     * @param view root for view
     */
    private void initView(View view) {
        lineChart = view.findViewById(R.id.line_chart_statistic);
    }

    /**
     * this method for initialize data
     */
    private void initLineChart() {

        lineChart.setOnChartValueSelectedListener(this);
        lineChart.getDescription().setEnabled(false);
        lineChart.setDrawBorders(true);
        lineChart.getAxisLeft().setEnabled(false);
        lineChart.getAxisRight().setDrawAxisLine(false);
        lineChart.getAxisRight().setDrawGridLines(false);
        lineChart.getXAxis().setDrawAxisLine(false);
        lineChart.getXAxis().setDrawGridLines(false);

        // no description text
        lineChart.setNoDataText("No data statistic for this psi");

        // enable value highlighting
        lineChart.setHighlightPerDragEnabled(true);
        lineChart.setHighlightPerTapEnabled(true);

        // enable touch gestures
        lineChart.setTouchEnabled(true);

        setLineChartScalingAndDragging();

        // if disabled, scaling can be done on x- and y-axis separately
        lineChart.setPinchZoom(true);

        setLegend();
    }

    private void setLineChartScalingAndDragging() {
        // enable scaling and dragging
        lineChart.setDragEnabled(true);
        lineChart.setScaleEnabled(true);
        lineChart.setDrawGridBackground(false);
    }

    private void setLegend() {
        Legend l = lineChart.getLegend();
        l.setVerticalAlignment(Legend.LegendVerticalAlignment.TOP);
        l.setHorizontalAlignment(Legend.LegendHorizontalAlignment.RIGHT);
        l.setOrientation(Legend.LegendOrientation.VERTICAL);
        l.setDrawInside(false);
    }

    /**
     * this method for set all data to line chart
     */
    @SuppressLint("ResourceType")
    private void setData() {
        psiByDate = PSIDatabaseImplementation.getPsiDate();

        for (int i = 0; i < psiByDate.getItemsArrayList().size(); i++) {
            addItemData(i);
        }

        setLineDataSet(R.color.colorPurple, arrayCentral, getLabel(R.string.string_central));
        setLineDataSet(R.color.colorGreen, arrayNorth, getLabel(R.string.string_north));
        setLineDataSet(R.color.colorOrange, arrayWest, getLabel(R.string.string_west));
        setLineDataSet(R.color.colorRed, arraySouth, getLabel(R.string.string_south));
        setLineDataSet(R.color.colorBlue, arrayEast, getLabel(R.string.string_east));
        setLineDataSet(R.color.colorYellow, arrayNational, getLabel(R.string.string_national));

        // create a data object with the datasets
        LineData data = new LineData(dataSets);

        // set data
        lineChart.setData(data);

        lineChart.animateX(2500);
        lineChart.invalidate();

        // get the legend (only possible after setting data)
        Legend l = lineChart.getLegend();

        // modify the legend ...
        l.setForm(Legend.LegendForm.LINE);
        l.setTextColor(R.color.colorGray);

        setAxisFormatter();

    }

    private void addItemData(int i) {
        arrayTime.add(ContentHelper.getDate("hh:mm:ss", psiByDate.getItemsArrayList().get(i).getTimestamp().substring(11, 19), "KK a"));
        float psiCentral = (float) getDataIndex(i).getPsiTwentyFourHourly().getCentral();
        float psiNorth = (float) getDataIndex(i).getPsiTwentyFourHourly().getNorth();
        float psiWest = (float) getDataIndex(i).getPsiTwentyFourHourly().getWest();
        float psiSouth = (float) getDataIndex(i).getPsiTwentyFourHourly().getSouth();
        float psiEast = (float) getDataIndex(i).getPsiTwentyFourHourly().getEast();
        float psiNational = (float) getDataIndex(i).getPsiTwentyFourHourly().getNational();

        arrayCentral.add(new Entry(i, psiCentral));
        arrayNorth.add(new Entry(i, psiNorth));
        arrayWest.add(new Entry(i, psiWest));
        arraySouth.add(new Entry(i, psiSouth));
        arrayEast.add(new Entry(i, psiEast));
        arrayNational.add(new Entry(i, psiNational));
    }

    private DataIndex getDataIndex(int i) {
        return psiByDate.getItemsArrayList().get(i).getDataIndex();
    }

    private String getLabel(@StringRes int label) {
        return getString(R.string.string_psi24h) + " " + getString(label);
    }

    @Override
    public void onValueSelected(Entry e, Highlight h) {

    }

    @Override
    public void onNothingSelected() {

    }

    private void setAxisFormatter() {
        IAxisValueFormatter formatter = new IAxisValueFormatter() {

            @Override
            public String getFormattedValue(float value, AxisBase axis) {
                return arrayTime.get((int) value);
            }

        };

        XAxis xAxis = lineChart.getXAxis();
        xAxis.setGranularity(1f); // minimum axis-step (interval) is 1
        xAxis.setValueFormatter(formatter);
    }

    private void setLineDataSet(int color, ArrayList<Entry> entryArrayList, String label) {
        // create a dataset and give it a type
        LineDataSet lineDataSet = new LineDataSet(entryArrayList, label);
        lineDataSet.setColor(getResources().getColor(color));
        lineDataSet.setCircleColor(color);
        lineDataSet.setLineWidth(2f);
        lineDataSet.setCircleSize(4f);
        lineDataSet.setFillAlpha(65);
        lineDataSet.setFillColor(color);
        lineDataSet.setHighLightColor(Color.rgb(244, 117, 117));
        dataSets.add(lineDataSet); // add the datasets
    }

}

File: app/src/main/java/com/imdglobalservices/psi/fragment/PSIStatisticFragment.kt
package com.imdglobalservices.psi.fragment

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.IAxisValueFormatter
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import com.github.mikephil.charting.listener.OnChartValueSelectedListener
import com.imdglobalservices.psi.R
import com.imdglobalservices.psi.database.PSIDatabaseImplementation
import com.imdglobalservices.psi.models.DataIndex
import com.imdglobalservices.psi.models.PSIDate
import com.imdglobalservices.psi.utils.ContentHelper
import kotlinx.android.synthetic.main.fragment_statistic_psi.*
import java.util.*

/**
 * @author Dwi Setiyono <dwi.setiyono></dwi.setiyono>@ovo.id>
 * @since 2017.16.11
 */
class PSIStatisticFragment : Fragment(), OnChartValueSelectedListener {
    private val arrayTime = ArrayList<String>()
    private val arrayCentral = ArrayList<Entry>()
    private val arrayNorth = ArrayList<Entry>()
    private val arrayWest = ArrayList<Entry>()
    private val arraySouth = ArrayList<Entry>()
    private val arrayEast = ArrayList<Entry>()
    private val arrayNational = ArrayList<Entry>()
    private val dataSets = ArrayList<ILineDataSet>()
    private var psiByDate: PSIDate? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater?.inflate(R.layout.fragment_statistic_psi, container, false)
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initline_chart_statistic()
        setData()
    }

    /**
     * this method for initialize data
     */
    private fun initline_chart_statistic() {

        line_chart_statistic?.setOnChartValueSelectedListener(this)
        line_chart_statistic?.description?.isEnabled = false
        line_chart_statistic?.setDrawBorders(true)
        line_chart_statistic?.axisLeft?.isEnabled = false
        line_chart_statistic?.axisRight?.setDrawAxisLine(false)
        line_chart_statistic?.axisRight?.setDrawGridLines(false)
        line_chart_statistic?.xAxis?.setDrawAxisLine(false)
        line_chart_statistic?.xAxis?.setDrawGridLines(false)

        // no description text
        line_chart_statistic?.setNoDataText("No data statistic for this psi")

        // enable value highlighting
        line_chart_statistic?.isHighlightPerDragEnabled = true
        line_chart_statistic?.isHighlightPerTapEnabled = true

        // enable touch gestures
        line_chart_statistic?.setTouchEnabled(true)

        setline_chart_statisticScalingAndDragging()

        // if disabled, scaling can be done on x- and y-axis separately
        line_chart_statistic?.setPinchZoom(true)

        setLegend()
    }

    private fun setline_chart_statisticScalingAndDragging() {
        // enable scaling and dragging
        line_chart_statistic?.isDragEnabled = true
        line_chart_statistic?.setScaleEnabled(true)
        line_chart_statistic?.setDrawGridBackground(false)
    }

    private fun setLegend() {
        val l = line_chart_statistic?.legend
        l?.verticalAlignment = Legend.LegendVerticalAlignment.TOP
        l?.horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
        l?.orientation = Legend.LegendOrientation.VERTICAL
        l?.setDrawInside(false)
    }

    /**
     * this method for set all data to line chart
     */
    @SuppressLint("ResourceType")
    private fun setData() {
        psiByDate = PSIDatabaseImplementation.psiDate

        for (i in 0 until psiByDate?.itemsArrayList!!.size) {
            addItemData(i)
        }

        setLineDataSet(R.color.colorPurple, arrayCentral, getLabel(R.string.string_central))
        setLineDataSet(R.color.colorGreen, arrayNorth, getLabel(R.string.string_north))
        setLineDataSet(R.color.colorOrange, arrayWest, getLabel(R.string.string_west))
        setLineDataSet(R.color.colorRed, arraySouth, getLabel(R.string.string_south))
        setLineDataSet(R.color.colorBlue, arrayEast, getLabel(R.string.string_east))
        setLineDataSet(R.color.colorYellow, arrayNational, getLabel(R.string.string_national))

        // create a data object with the datasets
        val data = LineData(dataSets)

        // set data
        line_chart_statistic?.data = data

        line_chart_statistic?.animateX(2500)
        line_chart_statistic?.invalidate()

        // get the legend (only possible after setting data)
        val l = line_chart_statistic?.legend

        // modify the legend ...
        l?.form = Legend.LegendForm.LINE
        l?.textColor = R.color.colorGray

        setAxisFormatter()

    }

    private fun addItemData(i: Int) {
        psiByDate?.itemsArrayList?.get(i)?.timestamp?.substring(11, 19)?.let { ContentHelper.getDate("hh:mm:ss", it, "KK a") }?.let { arrayTime.add(it) }
        val psiCentral = getDataIndex(i)?.psiTwentyFourHourly?.central!!.toFloat()
        val psiNorth = getDataIndex(i)?.psiTwentyFourHourly?.north!!.toFloat()
        val psiWest = getDataIndex(i)?.psiTwentyFourHourly?.west!!.toFloat()
        val psiSouth = getDataIndex(i)?.psiTwentyFourHourly?.south!!.toFloat()
        val psiEast = getDataIndex(i)?.psiTwentyFourHourly?.east!!.toFloat()
        val psiNational = getDataIndex(i)?.psiTwentyFourHourly?.national!!.toFloat()

        arrayCentral.add(Entry(i.toFloat(), psiCentral))
        arrayNorth.add(Entry(i.toFloat(), psiNorth))
        arrayWest.add(Entry(i.toFloat(), psiWest))
        arraySouth.add(Entry(i.toFloat(), psiSouth))
        arrayEast.add(Entry(i.toFloat(), psiEast))
        arrayNational.add(Entry(i.toFloat(), psiNational))
    }

    private fun getDataIndex(i: Int): DataIndex? {
        return psiByDate?.itemsArrayList?.get(i)?.dataIndex
    }

    private fun getLabel(@StringRes label: Int): String {
        return getString(R.string.string_psi24h) + " " + getString(label)
    }

    override fun onValueSelected(e: Entry, h: Highlight) {

    }

    override fun onNothingSelected() {

    }

    private fun setAxisFormatter() {
        val formatter = IAxisValueFormatter { value, axis -> arrayTime[value.toInt()] }

        val xAxis = line_chart_statistic?.xAxis
        xAxis?.granularity = 1f // minimum axis-step (interval) is 1
        xAxis?.valueFormatter = formatter
    }

    private fun setLineDataSet(color: Int, entryArrayList: ArrayList<Entry>, label: String) {
        // create a dataset and give it a type
        val lineDataSet = LineDataSet(entryArrayList, label)
        lineDataSet.color = resources.getColor(color)
        lineDataSet.setCircleColor(color)
        lineDataSet.lineWidth = 2f
        lineDataSet.circleSize = 4f
        lineDataSet.fillAlpha = 65
        lineDataSet.fillColor = color
        lineDataSet.highLightColor = Color.rgb(244, 117, 117)
        dataSets.add(lineDataSet) // add the datasets
    }

}
